File: Provenance/Controller/iCade/PViCadeGamepadButtonInput.m
//
//  PViCadeGamepadInputButton.m
//  Provenance
//
//  Created by Josejulio Martínez on 19/06/15.
//  Copyright (c) 2015 Josejulio Martínez. All rights reserved.
//

#import "PViCadeGamepadButtonInput.h"

@implementation PViCadeGamepadButtonInput

-(void) setValueChangedHandler:(GCControllerButtonValueChangedHandler) handler {
    _handler = handler;
}

-(void) setPressedChangedHandler:(GCControllerButtonValueChangedHandler) handler {
    _handler = handler;
}

-(void) buttonPressed {
    if (_handler) {
        _handler(self, 1.0, YES);
    }
    [self setPressed:YES];
}

-(void) buttonReleased {
    if (_handler) {
        _handler(self, 0.0, NO);
    }
    [self setPressed:NO];
}

-(void) setPressed:(BOOL)pressed {
    _pressed = pressed;
    _value = pressed ? 1.0f : 0.0f;
}

-(BOOL) isPressed {
    return _pressed;
}

- (float)value {
    return _value;
}

@end

File: PVSupport/Sources/PVSupport/Controller/iCade/PViCadeGamepadButtonInput.swift
//
//  PViCadeGamepadInputButton.swift
//  Provenance
//
//  Created by Joseph Mattiello
//  Copyright (c) 2022 Joseph Mattiello. All rights reserved.
//

#if canImport(UIKit) && canImport(GameController)
import GameController

public final class PViCadeGamepadButtonInput: GCControllerButtonInput {
    var handler: GCControllerButtonValueChangedHandler?

    private var _value: Float = 0.0
    public override var value: Float {
        get {
            return _value
        }
        @objc(setButtonValue:)
        set {
            _value = newValue
        }
    }

    private var _isPressed: Bool = false
    public override var isPressed: Bool {
        get {
            return _isPressed
        }
        set {
            _isPressed = newValue
            _value = _isPressed ? 1.0 : 0.0
            handler?(self, _value, _isPressed)
        }
    }

    public override var valueChangedHandler: GCControllerButtonValueChangedHandler? {
        get {
            return super.valueChangedHandler
        }
        set(handler) {
            self.handler = handler
        }
    }

    public override var pressedChangedHandler: GCControllerButtonValueChangedHandler? {
        get {
            return super.pressedChangedHandler
        }
        set(handler) {
            self.handler = handler
        }
    }
}
#endif
